opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java

```diff
@@ -0,0 +1,475 @@
+    final OpenTelemetryMetricsModule module;
+    final StreamInfo info;
+    final String fullMethodName;
+    volatile long outboundWireSize;
+    volatile long inboundWireSize;
+    long attemptNanos;
+    Code statusCode;
+
+    ClientTracer(CallAttemptsTracerFactory attemptsState, OpenTelemetryMetricsModule module,
+        StreamInfo info, String fullMethodName) {
+      this.attemptsState = attemptsState;
+      this.module = module;
+      this.info = info;
+      this.fullMethodName = fullMethodName;
+      this.stopwatch = module.stopwatchSupplier.get().start();
+    }
+
+    @Override
+    @SuppressWarnings("NonAtomicVolatileUpdate")
+    public void outboundWireSize(long bytes) {
+      if (outboundWireSizeUpdater != null) {
+        outboundWireSizeUpdater.getAndAdd(this, bytes);
+      } else {
+        outboundWireSize += bytes;
+      }
+    }
+
+    @Override
+    @SuppressWarnings("NonAtomicVolatileUpdate")
+    public void inboundWireSize(long bytes) {
+      if (inboundWireSizeUpdater != null) {
+        inboundWireSizeUpdater.getAndAdd(this, bytes);
+      } else {
+        inboundWireSize += bytes;
+      }
+    }
+
+    @Override
+    @SuppressWarnings("NonAtomicVolatileUpdate")
+    public void inboundMessage(int seqNo) {
+      if (inboundReceivedOrClosed.compareAndSet(false, true)) {
+        // Because inboundUncompressedSize() might be called after streamClosed(),
+        // we will report stats in callEnded(). Note that this attempt is already committed.
+        attemptsState.inboundMetricTracer = this;
+      }
+    }
+
+
+    @Override
+    public void streamClosed(Status status) {
+      stopwatch.stop();
+      attemptNanos = stopwatch.elapsed(TimeUnit.NANOSECONDS);
+      Deadline deadline = info.getCallOptions().getDeadline();
+      statusCode = status.getCode();
+      if (statusCode == Code.CANCELLED && deadline != null) {
+        // When the server's deadline expires, it can only reset the stream with CANCEL and no
+        // description. Since our timer may be delayed in firing, we double-check the deadline and
+        // turn the failure into the likely more helpful DEADLINE_EXCEEDED status.
+        if (deadline.isExpired()) {
+          statusCode = Code.DEADLINE_EXCEEDED;
+        }
+      }
+      attemptsState.attemptEnded();
+      if (inboundReceivedOrClosed.compareAndSet(false, true)) {
+        // Stream is closed early. So no need to record metrics for any inbound events after this
+        // point.
+        recordFinishedAttempt();
+      } // Otherwise will report metrics in callEnded() to guarantee all inbound metrics are
+      // recorded.
+    }
+
+    void recordFinishedAttempt() {
+      // TODO(dnvindhya) : add target as an attribute
+      io.opentelemetry.api.common.Attributes attribute = io.opentelemetry.api.common.Attributes.of(
+          AttributeKey.stringKey(OpenTelemetryConstants.METHOD_KEY), fullMethodName,
```

Comment:
Is there a reason this isn't an AttributeKey in OpenTelemetryConstants? It seems all the usages convert it into a key.